core/src/transport/circuit_breaking.rs

```diff
@@ -0,0 +1,383 @@
+                        // Avoid open circuits for unsupported addresses.
+                        return Err(TransportError::MultiaddrNotSupported(err));
+                    }
+                    Err(err) => {
+                        self.state.open_or_extend_circuit(addr.clone());
+                        Err(err)
+                    }
+                };
+                CircuitBreakingDialInner::Dialing(future?)
+            }
+
+            Some(wait_duration) => {
+                // Wait before dialing.
+                let timer = Delay::new(wait_duration);
+                CircuitBreakingDialInner::Waiting(timer, Some(self.inner))
+            }
+        };
+
+        Ok(CircuitBreakingDial {
+            inner,
+            addr,
+            state: self.state,
+        })
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner.address_translation(listen, observed)
+    }
+}
+
+/// Wraps around a `Future` that produces a connection. Adds a delay before
+/// dialing the address if its circuit is opened.
+#[pin_project]
+pub struct CircuitBreakingDial<TInner>
+where
+    TInner: Transport,
+{
+    /// The inner state of the future.
+    #[pin]
+    inner: CircuitBreakingDialInner<TInner>,
+    /// The address to be dialed.
+    addr: Multiaddr,
+    /// The shared state containing information about circuits.
+    state: Arc<CircuitBreakerState>,
+}
+
+#[pin_project(project = CircuitBreakingDialInnerProj)]
+enum CircuitBreakingDialInner<TInner>
+where
+    TInner: Transport,
+{
+    /// Due to an open circuit the address is diales only after `Delay`.
+    Waiting(#[pin] Delay, Option<TInner>),
+    /// The address is currently being dialed.
+    Dialing(#[pin] TInner::Dial),
+}
+
+impl<TInner> Future for CircuitBreakingDial<TInner>
+where
+    TInner: Transport,
+{
+    type Output = Result<TInner::Output, TInner::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        loop {
+            match this.inner.as_mut().project() {
+                CircuitBreakingDialInnerProj::Waiting(delay, transport) => {
+                    match delay.poll(cx) {
+                        Poll::Pending => return Poll::Pending,
+                        Poll::Ready(()) => {
+                            let transport = transport
+                                .take()
+                                .expect("future called after being finished");
+                            match transport.dial(this.addr.clone()) {
+                                Ok(f) => this.inner.set(CircuitBreakingDialInner::Dialing(f)),
+                                Err(TransportError::MultiaddrNotSupported(_addr)) => {
+                                    // TODO do something other than panic?
+                                    panic!("Open circuit for unsupported Multiaddr")
```

Comment:
I assumed this arm is not supposed to be executed and therefore it's ok to `panic!`. Not sure, though. Also there might be a more elegant solution.

Reply:
Instead of delaying the call to `TInner::dial`, one could do the call in `CircuitBreaking::dial` directly, only delaying the polling of the `TInner::Dial`, not the creation. That of course assumes that `Tinner::Dial` does nothing without being polled. I am not sure whether all implementations of `Transport` uphold this assumption.

Reply:
To verify it's probably required to check all `Transport` implementations (and assume they will continue to uphold this assumption in the future).

Could a `Multiaddr` which was initially supported when the circuit was opened become unspupported while waiting for the circuit to close?

Reply:
> Could a Multiaddr which was initially supported when the circuit was opened become unspupported while waiting for the circuit to close?

In theory yes. If we ever decide this to be invalid, we should change the `TransportError` type to prevent his at type level.

For now, I would suggest keeping it as is, adding a comment on `fn dial` above documenting the potential panic.

Reply:
Addressed in 93fec82 .